qualtran/bloqs/chemistry/trotter.ipynb

```diff
@@ -0,0 +1,752 @@
+    "        rsq_one.append(rsq)\n",
+    "    elif rsq in bins_two[bl]:\n",
+    "        poly_two.append(polynomial_approx_range_two_scaling(rsq, scale_power=bl))\n",
+    "        rsq_two.append(rsq)\n",
+    "    else:\n",
+    "        raise ValueError(\"Could not find bin for rsq.\")\n",
+    "\n",
+    "plt.plot(rsq_one, poly_one, marker='o', lw=0, label='range one')\n",
+    "plt.plot(rsq_two, poly_two, marker='x', lw=0, label='range two')\n",
+    "plt.plot(rij_int**2, 1/rij_int, label='Expected')\n",
+    "plt.legend()\n",
+    "plt.xlabel(\"$r_{ij}^2$\")\n",
+    "plt.ylabel(r\"$\\frac{1}{\\sqrt{r_{ij}^2}}$\")\n",
+    "# Check the global error\n",
+    "combined_r = np.concatenate([rsq_one, rsq_two])\n",
+    "ix = np.argsort(combined_r)\n",
+    "combined = np.concatenate([poly_one, poly_two])[ix]\n",
+    "nr_one = newton_raphson_step(np.array(rsq_one), np.array(poly_one), delta_range_one)\n",
+    "nr_two = newton_raphson_step(np.array(rsq_two), np.array(poly_two), delta_range_two)\n",
+    "combined_nr = np.concatenate([nr_one, nr_two])[ix]\n",
+    "print(f\"max error = {np.max(np.abs(combined - 1/rij_int))} vs expected error = {2**(-15)}\")\n",
+    "print(f\"max error after NR step = {np.max(np.abs(combined_nr - 1/rij_int))}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Ok! You might be wondering what is special about the intervals we chose to bin our integer $r_{ij}^2$ values by, and how QROM is connected with this. We'll try to explain that now.\n",
+    "\n",
+    "### ```QROM```\n",
+    "Recall that ultimately we need to load our polynomial coefficients onto the quantum computer before we can evaluate the inverse square root. Data loading can be achieved through QROM which implements: \n",
+    "\n",
+    "$$\n",
+    "\\mathrm{QROM}_d \\sum_l \\alpha_l |l\\rangle|0\\rangle = \\sum_l \\alpha_l |l\\rangle|d_l\\rangle.\n",
+    "$$\n",
+    "\n",
+    "That is, given a selection register $l$, QROM can load the binary representation of the $l$-th data element of $d$ into a target register of a given size.\n",
+    "In our case $|r_{ij}^2\\rangle$ is the selection register of size $2 n + 2$, and we want to load $b$-bit binary representations of our polynomial coefficients $\\{a_0, a_1, a_2, a_3\\}$ and $\\{b_0, b_1, b_2, b_3\\}$ for the two ranges.\n",
+    "\n",
+    "In principle we could load all of the (scaled) coefficients for each value of $r_{ij}$ in the register, but this would incur a cost that goes like $L-1$ Toffolis, where $L=2^{2n + 2}$ which would be unacceptably large. But we just saw that much of this data would be duplicated as we only need the coefficients for each subinterval, of which there are a logarithmic amount. Fortunately we can exploit the structure of the unary iteration circuit to exploit this redundancy and significantly reduce our costs and arrive at `variably space QROM`.\n",
```

Comment:
lol why do you put backticks around `variable space QROM`

Reply:
Ah, I am hoping these will become links to notebooks on the topics in the future. But will remove.

Reply:
In markdown, links are `[put in brackets](www.url.com)` but we shouldn't put in dummy links to nowhere